Add on_cleared event

PhrozenByte opened the pull request:
This event is fired when a notification was closed without user interaction, e.g. because the notification timed out (DBus only, and only if supported by the notifications server; undetectable by capabilities), or because the notification was closed by another process (DBus only).

We'll later also utilize this for the cross-platform implementation of `timeout`.

Without this new event we'd have to remove the `if reason == NOTIFICATION_CLOSED_DISMISSED:` check in `DBusDesktopNotifier`.

**ToDo:**
* [ ] After merging # (Quality of life improvements to Notification class) `Notification.on_cleared` must be changed to:
    ```python
        on_cleared: Callable[[], Any] | None = field(default=None, repr=False)
    ```
* [ ] After merging # (DesktopNotifierBackend: Drop `notification` param from `handle_*()` methods) `DesktopNotifierBackend.on_cleared` must be changed to:
    ```python
        def handle_cleared(self, identifier: str) -> None:
            notification = self._clear_notification_from_cache(identifier)
            # …
    ```
* [ ] After merging # (DesktopNotifierBackend: Drop `notification` param from `handle_*()` methods) `DesktopNotifierBackend.on_cleared` also update the docstring of `DesktopNotifier.on_cleared`
* [ ] Add `on_cleared` to `DesktopNotifierSync` after merging # (DesktopNotifierSync: Add `on_*` properties for class-level events)

samschott commented:
As for https://github.com/samschott/desktop-notifier/pull/209, could you check if similar events are available on macOS and Windows?

PhrozenByte commented:
Windows doesn't have a similar event (I checked).

Can't say anything about OS X because I neither understand their overly complex API, nor do I have an Apple device to do manual testing. I'm afraid anything OS-X-related is up to you.

As said earlier, my ultimate goal is to use this for #206 too, i.e. independent of the backend used. I'm just not sure yet how to prevent calling the dismissal class-level handler. However, this isn't really relevant here, see #206.

samschott commented on src/desktop_notifier/common.py:
```diff
@@ -289,6 +292,10 @@ class Capability(Enum):
+    ON_CLEARED = auto()
+    """Supports distinguishing between an user closing a notification, and clearing a
+    notification programmatically, and consequently supports on-cleared callbacks"""
+
```
How about "Supports callbacks on programmatic closing and notification timeouts"?

PhrozenByte replied:
Hmm… Reading through both my original suggestion five months later and yours now, I'd interpret both like "if this cap isn't set, the notification can't be closed programmatically". I made another attempt, WDYT?

samschott reviewed (commented):
Thanks and apologies for the delay in reviewing!

Do you think there is any value in distinguishing between expired and programatically closed notifications?

PhrozenByte commented:
> Do you think there is any value in distinguishing between expired and programatically closed notifications?

We could.

It's been five months, so I've forgotten a lot of my original reasoning :laughing: Sometimes that's a good thing: I now spent quite some time to think about this, especially in regards to #206, and distinguishing between those as well makes things easier for #206, too: We could then have a separate handler for #206 and the issues with #206 about duplicate events magically disappear thanks to `DispatchedNotification`. So, I definitely like your idea :+1:

I've rebased and updated this PRs accordingly. I'll update the other PRs as well.

samschott commented:
@PhrozenByte, did you actually update this PR to separate timeout and programmatic clearing calbacks?

PhrozenByte commented:
Yes. I assume you ask because I forgot to remove `timeout` from the examples? I just fixed that, sorry. Since I was over it I also now fully unified the examples. Should be ready-to-merge now :+1:
